src/arch/src/lib.rs

```diff
@@ -63,43 +61,88 @@ impl From<GuestMemoryError> for Error {
 
 pub type Result<T> = std::result::Result<T, Error>;
 
-pub fn create_fdt<T: GuestMemory>(
-    cmdline: &str,
-    num_vcpus: u32,
-    guest_mem: &T,
-    fdt_load_offset: u64,
-    serial_addr: u64,
-    rtc_addr: u64,
-) -> Result<()> {
-    let mut fdt = FdtWriter::new()?;
-
-    // The whole thing is put into one giant node with s
-    // ome top level properties
-    let root_node = fdt.begin_node("")?;
-    fdt.property_u32("interrupt-parent", PHANDLE_GIC)?;
-    fdt.property_string("compatible", "linux,dummy-virt")?;
-    fdt.property_u32("#address-cells", 0x2)?;
-    fdt.property_u32("#size-cells", 0x2)?;
-
-    create_chosen_node(&mut fdt, cmdline)?;
-
-    let mem_size: u64 = guest_mem.iter().map(|region| region.len() as u64).sum();
-    create_memory_node(&mut fdt, mem_size)?;
-    create_cpu_nodes(&mut fdt, num_vcpus)?;
-    create_gic_node(&mut fdt, true, num_vcpus as u64)?;
-    create_serial_node(&mut fdt, serial_addr)?;
-    create_rtc_node(&mut fdt, rtc_addr)?;
-    create_timer_node(&mut fdt, num_vcpus)?;
-    create_psci_node(&mut fdt)?;
-    create_pmu_node(&mut fdt, num_vcpus)?;
-
-    fdt.end_node(root_node)?;
-    let fdt_final = fdt.finish()?;
-
-    let fdt_address = GuestAddress(AARCH64_PHYS_MEM_START + fdt_load_offset);
-    guest_mem.write_slice(fdt_final.as_slice(), fdt_address)?;
+#[derive(Default)]
+pub struct FdtBuilder<'a> {
+    cmdline: Option<&'a str>,
+    mem_size: Option<u64>,
+    num_vcpus: Option<u32>,
+    serial_console: Option<(u64, u64)>,
+    rtc: Option<(u64, u64)>,
+}
 
-    Ok(())
+impl<'a> FdtBuilder<'a> {
+    pub fn new() -> Self {
+        FdtBuilder::default()
+    }
+
+    pub fn with_cmdline(&mut self, cmdline: &'a str) -> &mut Self {
+        self.cmdline = Some(cmdline);
+        self
+    }
+
+    pub fn with_mem_size(&mut self, mem_size: u64) -> &mut Self {
+        self.mem_size = Some(mem_size);
+        self
+    }
+
+    pub fn with_num_vcpus(&mut self, num_vcpus: u32) -> &mut Self {
+        self.num_vcpus = Some(num_vcpus);
+        self
+    }
+
+    pub fn with_serial_console(&mut self, addr: u64, size: u64) -> &mut Self {
+        self.serial_console = Some((addr, size));
+        self
+    }
+
+    pub fn with_rtc(&mut self, addr: u64, size: u64) -> &mut Self {
+        self.rtc = Some((addr, size));
+        self
+    }
+
+    pub fn create_fdt(&self) -> Result<Fdt> {
+        let mut fdt = FdtWriter::new()?;
+
+        // The whole thing is put into one giant node with s
+        // ome top level properties
+        let root_node = fdt.begin_node("")?;
+        fdt.property_u32("interrupt-parent", PHANDLE_GIC)?;
+        fdt.property_string("compatible", "linux,dummy-virt")?;
+        fdt.property_u32("#address-cells", 0x2)?;
+        fdt.property_u32("#size-cells", 0x2)?;
+
+        create_chosen_node(&mut fdt, self.cmdline.unwrap_or_default())?;
```

Comment:
The cmdline, num_vcpus, and mem_size should be mandatory. if these are not specified, than we should return an error.

Reply:
If these are mandatory, should we mark them as such in the FdtBuilder? Looks a bit unusual to have all the struct members as optional.

Reply:
If we don't use Option, we need to set a default value to them. And there doesn't seem to be a suitable default value to set to them

Reply:
I was thinking to pass the required parameters to `new`, similar to how it is described [here](https://doc.rust-lang.org/1.0.0/style/ownership/builders.html).

Reply:
What do you think? @andreeaflorescu

Reply:
I think it's better to not pass them in new because then the new function will have too many parameters. I prefer to keep them as they are now. It is typical for the builder to have all parameters as Optional because that's how you get to pass them one by one. Parameters that are required can only be verified at the end, when the object we want to build is done with the configuration.

Reply:
When you only have one parameter that's not optional, then it's fine to pass it to the `new` function, especially if it's something obvious.

Reply:
:+1: